### PR TITLE
tstest/natlab: correctly handle dual-stacked PacketConns.

### PR DESCRIPTION
**Pull request by danderson**
Adds a test with multiple networks, one of which is v4-only.

---
This change is [Reviewable](https://reviewable.io/reviews/tailscale/tailscale/518)

### CODE REVIEW

File: tstest/natlab/natlab_test.go [hunk 1]
```diff
@@ -77,45 +77,68 @@ func TestSendPacket(t *testing.T) {
 	}
 }
 
-func TestLAN(t *testing.T) {
-	// TODO: very duplicate-ey with the previous test, but important
-	// right now to test explicit construction of Networks.
+func TestMultiNetwork(t *testing.T) {
 	lan := Network{
```
Comment (bradfitz) [inline tstest/natlab/natlab_test.go:81]: style nit: `&Network` (since it's used as a pointer receiver)

File: tstest/natlab/natlab_test.go [hunk 2]
```diff
@@ -77,45 +77,68 @@ func TestSendPacket(t *testing.T) {
...
 
-	foo := NewMachine("foo")
-	bar := NewMachine("bar")
-	ifFoo := foo.Attach("eth0", &lan)
-	ifBar := bar.Attach("eth0", &lan)
+	client := NewMachine("client")
+	nat := NewMachine("nat")
+	server := NewMachine("server")
+
+	ifClient := client.Attach("eth0", &lan)
```
Comment (bradfitz) [inline tstest/natlab/natlab_test.go:91]: then you can remove all these `&`

File: tstest/natlab/natlab.go [hunk 3]
```diff
@@ -184,27 +184,33 @@ type Machine struct {
 	interfaces []*Interface
 	routes     []routeEntry // sorted by longest prefix to shortest
 
-	conns map[netaddr.IPPort]*conn
+	conns4 map[netaddr.IPPort]*conn
```
Comment (bradfitz) [inline tstest/natlab/natlab.go:187]: comments on these?

File: tstest/natlab/natlab.go [hunk 4]
```diff
@@ -184,27 +184,33 @@ type Machine struct {
...
+	for _, dst := range possibleDsts {
+		c, ok := conns[dst]
 		if !ok {
-			return
+			continue
+		}
+		select {
+		case c.in <- incomingPacket{src: src, p: p}:
+		default:
+			// Queue overflow. Just drop it.
```
Comment (bradfitz) [inline tstest/natlab/natlab.go:212]: I think you want to return now, not keep looping through `possibleDsts`?

File: tstest/natlab/natlab.go [hunk 5]
```diff
@@ -284,7 +290,12 @@ func (m *Machine) writePacket(p []byte, dst, src netaddr.IPPort) (n int, err err
 	case src.IP == v4unspec:
 		src.IP = iface.V4()
 	case src.IP == v6unspec:
-		src.IP = iface.V6()
+		// v6unspec in Go means "any src, but match address families"
```
Comment (bradfitz) [inline tstest/natlab/natlab.go:293]: nice fix

File: tstest/natlab/natlab.go [hunk 6]
```diff
@@ -321,59 +332,86 @@ func (m *Machine) hasv6() bool {
...
 }
 
-func (m *Machine) registerConn(c *conn) error {
+func (m *Machine) registerConn4(c *conn) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if c.ipp.IP.Is6() && c.ipp.IP != v6unspec {
+		return fmt.Errorf("registerConn4 got IPv6 %s", c.ipp)
+	}
+	if _, ok := m.conns4[c.ipp]; ok {
```
Comment (bradfitz) [inline tstest/natlab/natlab.go:341]: at least from here down can be shared with registerConn6 with a new helper `registerConn(m *map[net.IPPort]*conn, ipp net.IPPort, c *conn)`

File: tstest/natlab/natlab.go [hunk 7]
```diff
@@ -382,8 +420,22 @@ func (m *Machine) ListenPacket(network, address string) (net.PacketConn, error)
...
 		in:  make(chan incomingPacket, 100), // arbitrary
 	}
-	if err := m.registerConn(c); err != nil {
-		return nil, err
+	switch c.fam {
+	case 0:
+		if err := m.registerConn4(c); err != nil {
+			return nil, err
+		}
+		if err := m.registerConn6(c); err != nil {
```
Comment (bradfitz) [inline tstest/natlab/natlab.go:428]: if this fails, unregisterConn4?

File: tstest/natlab/natlab.go (hunk 4 above)
Comment (danderson) [reply tstest/natlab/natlab.go:212]: Fixed incidentally in followup change.

**Pull request merged**